applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+# if CONFIG_UART_LINE_CTRL
+			cdc_dtr = 0;
+# else
+			cdc_dtr = 1;
+# endif
+#endif
+
+			initialized = true;
+
+			k_sem_give(&sem_initialized);
+		}
+
+		return false;
+	}
+
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			const struct config_event *event;
+
+			event = cast_config_event(eh);
+
+			if (!is_my_config_id(event->id)) {
+				return false;
+			}
+
+			handle_config_event(event);
+
+			return false;
+		}
+
+		if (is_config_fetch_request_event(eh)) {
+			const struct config_fetch_request_event *event =
+				cast_config_fetch_request_event(eh);
+
+			if (is_my_config_id(event->id)) {
+				struct config_fetch_event *fetch_event;
+
+				switch (OPT_FIELD_GET(event->id)) {
+				case QOS_OPT_BLACKLIST:
+					fetch_event =
+						fetch_event_qos_blacklist();
+					break;
+				case QOS_OPT_CHMAP:
+					fetch_event =
+						fetch_event_qos_channel_map();
+					break;
+				case QOS_OPT_PARAM_BLE:
+					fetch_event =
+						fetch_event_qos_ble_params();
+					break;
+				case QOS_OPT_PARAM_WIFI:
+					fetch_event =
+						fetch_event_qos_wifi_params();
+					break;
+				default:
+					LOG_WRN("Unknown opt");
+					return false;
+				}
+
+				fetch_event->id = event->id;
+				fetch_event->recipient = event->recipient;
+				fetch_event->channel_id = event->channel_id;
+
+				EVENT_SUBMIT(fetch_event);
+			}
+
+			return false;
+		}
+	}
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+	if (IS_ENABLED(CONFIG_DESKTOP_HID_MOUSE)) {
+		static s32_t m_hid_pkt_recv_count;
+		static s32_t m_usb_pkt_sent_count;
+		static u32_t m_cdc_notify_count;
+
+		/* Count number of HID packets received via BLE */
+		/* and number of HID packets sent via USB. */
+		/* Send stats printout via CDC every 100 packets */
+
+		if (is_hid_mouse_event(eh)) {
+			++m_hid_pkt_recv_count;
+			++m_cdc_notify_count;
+
+			if (m_cdc_notify_count == 100) {
+				hid_pkt_stats_print(
+					m_hid_pkt_recv_count,
+					m_usb_pkt_sent_count);
+				m_cdc_notify_count = 0;
+			}
+
+			return false;
+		}
+
+		if (is_hid_report_sent_event(eh)) {
+			++m_usb_pkt_sent_count;
+
+			return false;
+		}
+	}
+#endif
+
+	/* If event is unhandled, unsubscribe. */
+	__ASSERT_NO_MSG(false);
+
+	return false;
+}
+
+static void apply_new_params(void)
+{
+	int err;
+
+	k_mutex_lock(&m_data_access_mutex, K_FOREVER);
+	/* chmap_filter_params_set Returns immediately */
+	err = chmap_filter_params_set(mp_chmap_inst, m_new_params);
+	k_free(m_new_params);
+	m_new_params = NULL;
+	k_mutex_unlock(&m_data_access_mutex);
+
+	if (err) {
+		LOG_WRN("Param update failed");
+	}
+}
+
+void ble_qos_thread_fn(void)
+{
+	int err;
+
+	err = k_sem_take(&sem_initialized, K_FOREVER);
+	if (err) {
+		LOG_ERR("Init failed: %d", err);
+		__ASSERT_NO_MSG(false);
+	}
+
+	for (;;) {
+		bool update_channel_map;
+
+		k_sleep(1000); /* Set time unit for chmap filter to ~1 second */
+
+		/* Check and apply new parameters received via config channel */
+		if (m_new_params != NULL) {
```

Comment:
In this particular case it will work. The only downside can be when `new_params` will be updated while check is done which will delay the update.

Still I recommend to use lock everytime protected resource is touched. This ensures that data is consistent and required barriers are applied.
Fixing this would actually be very simple - just move the check to `apply_new_params`. E.g.:
```
static void apply_params(void)
{
   mutex_lock(&params_lock);
   if (params_update_request) { // better to have statically allocated memory to store new settings
      do_stuff();
      params_update_request = false;
   }
   LOG_DBG("some log"); // logging is assumed to be fast enough to be used in ISR so can go under lock
   mutex_unlock(&params_lock);
}
```